WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeViewModel.swift

```diff
@@ -0,0 +1,110 @@
+
+    var sections: [Section] = []
+
+    /// Represents the current state of `synchronizeProductAttributes` action. Useful for the consumer to update it's UI upon changes
+    ///
+    enum SyncingState: Equatable {
+        case initialized
+        case syncing
+        case failed
+        case synced
+    }
+
+    /// Closure to be invoked when `synchronizeProductAttributes` state changes
+    ///
+    private var onSyncStateChange: ((SyncingState) -> Void)?
+
+    /// Current  product attributes synchronization state
+    ///
+    private var syncProductAttributesState: SyncingState = .initialized {
+        didSet {
+            guard syncProductAttributesState != oldValue else {
+                return
+            }
+            onSyncStateChange?(syncProductAttributesState)
+        }
+    }
+
+    init(storesManager: StoresManager = ServiceLocator.stores, product: Product) {
+        self.storesManager = storesManager
+        self.product = product
+    }
+
+    /// Load existing product attributes from storage and fire the synchronize all product attributes action.
+    ///
+    func performFetch() {
+        synchronizeAllProductAttributes()
+    }
+
+    /// Observes and notifies of changes made to product attributes. The current state will be dispatched upon subscription.
+    /// Calling this method will remove any other previous observer.
+    ///
+    func observeProductAttributesListStateChanges(onStateChanges: @escaping (SyncingState) -> Void) {
+        onSyncStateChange = onStateChanges
+        onSyncStateChange?(syncProductAttributesState)
+    }
+}
+
+// MARK: - Synchronize global Product Attributes
+//
+private extension AddAttributeViewModel {
+    /// Synchronizes all global product attributes.
+    ///
+    func synchronizeAllProductAttributes() {
+        syncProductAttributesState = .syncing
+        let action = ProductAttributeAction.synchronizeProductAttributes(siteID: product.siteID) { [weak self] result in
+            switch result {
+            case .success(let attributes):
+                self?.updateSections(attributes: attributes)
+                self?.syncProductAttributesState = .synced
+            case .failure(let error):
+                DDLogError("⛔️ Error fetching product attributes: \(error.localizedDescription)")
+                self?.syncProductAttributesState = .failed
+            }
+        }
+        storesManager.dispatch(action)
+    }
+
+    /// Updates  data from the remote fetched objects.
+    ///
+    func updateSections(attributes: [ProductAttribute]) {
+
+        /// Sum fetched attributes + attributes inside the product (global + local), and remove duplicated product attributes, then sort it
+        localAndGlobalAttributes = attributes + product.attributes.filter { element in
+            return !attributes.contains { $0.attributeID == element.attributeID }
+        }.sorted(by: { (attribute1, attribute2) -> Bool in
```

Comment:
As far as I know, in theory this could be written as `sorted(by: > )`, I am not sure that makes it more readable though

Reply:
wow good suggestion, I implemented it 💯 8cf059f555614b9c1b47b8d207517780d4184d36